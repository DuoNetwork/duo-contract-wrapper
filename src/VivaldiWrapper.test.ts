// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import * as CST from './constants';
import { ITransactionOption } from './types';
import VivaldiWrapper from './VivaldiWrapper';

test('convertCustodianState', () => {
	expect(VivaldiWrapper.convertCustodianState(CST.STATE_INCEPTION)).toBe(CST.CTD_INCEPTION);
	expect(VivaldiWrapper.convertCustodianState(CST.STATE_PRERESET)).toBe(CST.CTD_PRERESET);
	expect(VivaldiWrapper.convertCustodianState(CST.STATE_RESET)).toBe(CST.CTD_RESET);
	expect(VivaldiWrapper.convertCustodianState(CST.STATE_TRADING)).toBe(CST.CTD_TRADING);
	expect(VivaldiWrapper.convertCustodianState(CST.STATE_MATURED)).toBe(CST.CTD_MATURED);
	expect(VivaldiWrapper.convertCustodianState('any')).toBe(CST.CTD_LOADING);
});

const collectFeeSend = jest.fn(() => Promise.resolve());
const setValueSend = jest.fn(() => Promise.resolve());
const startRoundSend = jest.fn(() => Promise.resolve());
const endRoundSend = jest.fn(() => Promise.resolve());
const forceEndRoundSend = jest.fn(() => Promise.resolve());
const createSend = jest.fn(() => {
	const on = jest.fn();
	setTimeout(() => on.mock.calls[0][1]('createTxHash'), 0);
	return {
		on: on
	};
});

const redeemSend = jest.fn(() => {
	const on = jest.fn();
	setTimeout(() => on.mock.calls[0][1]('redeemTxHash'), 0);
	return {
		on: on
	};
});
const startCustodianSend = jest.fn(() => Promise.resolve());
const startPreResetSend = jest.fn(() => Promise.resolve());
const startResetSend = jest.fn(() => Promise.resolve());

const web3Wrapper = {
	isReadOnly: jest.fn(() => true),
	onSwitchToMetaMask: jest.fn(() => ({} as any)),
	fromWei: jest.fn(value => value * 1e-18),
	toWei: jest.fn(value => value * 1e18),
	onSwitchToLedger: jest.fn(() => ({} as any)),
	readOnlyReject: jest.fn(() => Promise.reject('Read Only Mode')),
	wrongEnvReject: jest.fn(() => Promise.reject('wrong env')),
	getGasPrice: jest.fn(() => Promise.resolve(1000000000)),
	getTransactionCount: jest.fn(() => Promise.resolve(2)),
	getTransactionOption: jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	),
	createContract: jest.fn(() => ({
		methods: {
			isKnockedIn: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve(true))
			})),
			getStates: jest.fn(() => ({
				call: jest.fn(() =>
					Promise.resolve({
						'0': '12345678980',
						'1': '1000000000000000000000',
						'2': '100000000000000000',
						'3': '100000000000000000000000',
						'4': '100000',
						'5': '100',
						'6': '1000000000000000000',
						'7': '2000000000000000000',
						'8': '100000000000000000000',
						'9': '1000000000',
						'10': '100000000000000000000',
						'11': '10000000000000000000000000',
						'12': '1000000000000000000',
						'13': '1000000000000000000',
						'14': '1000000000000000000',
						'15': '1000000000000000000',
						'16': '1000000000000000000',
						'17': '1000000000000000000',
						'18': '1000000000000000000000',
						'19': '10000000000000000000',
						'20': '1000000000000000000000',
						'21': '100000000000000000000',
						'22': '1000000000000000000',
					})
				)
			})),
			getAddresses: jest.fn(() => ({
				call: jest.fn(() =>
					Promise.resolve({
						'0': 'roleManager',
						'1': 'operator',
						'2': 'feeCollector',
						'3': 'oracle',
						'4': 'aToken',
						'5': 'bToken',
						'6': 'collateralToken'
					})
				)
			})),
			users: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve({}))
			})),
			collectFee: jest.fn(() => ({
				send: collectFeeSend
			})),
			setValue: jest.fn(() => ({
				send: setValueSend
			})),
			startRound: jest.fn(() => ({
				send: startRoundSend
			})),
			endRound: jest.fn(() => ({
				send: endRoundSend
			})),
			forceEndRound: jest.fn(() => ({
				send: forceEndRoundSend
			})),
			create: jest.fn(() => ({
				send: createSend
			})),
			redeem: jest.fn(() => ({
				send: redeemSend
			})),
			startCustodian: jest.fn(() => ({
				send: startCustodianSend
			})),
			startReset: jest.fn(() => ({
				send: startResetSend
			})),
			startPreReset: jest.fn(() => ({
				send: startPreResetSend
			})),
			updateRoleManager: jest.fn(() => ({
				send: jest.fn()
			})),
			updateOperator: jest.fn(() => ({
				send: jest.fn()
			})),
			updateOracle: jest.fn(() => ({
				send: jest.fn()
			})),
			updateFeeCollector: jest.fn(() => ({
				send: jest.fn()
			}))
		}
	}))
} as any;

const vivaldiWrapper = new VivaldiWrapper(web3Wrapper, 'address');

test('getStates', async () => {
	expect(await vivaldiWrapper.getStates()).toMatchSnapshot();
});

test('getAddress', async () => {
	expect(await vivaldiWrapper.getAddresses()).toMatchSnapshot();
});

test('getUserAddress', async () => {
	await vivaldiWrapper.getUserAddress(1);
	expect((vivaldiWrapper.contract.methods.users as jest.Mock).mock.calls).toMatchSnapshot();
});

test('collectFee', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	try {
		await vivaldiWrapper.collectFee('account', 10);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}

	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.collectFee('account', 10);
	await vivaldiWrapper.collectFee('account', 10, {
		gasLimit: 20000,
		gasPrice: 2000000000,
		nonce: 40
	});
	expect((vivaldiWrapper.contract.methods.collectFee as jest.Mock).mock.calls).toMatchSnapshot();
	expect(collectFeeSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('setValue', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.setValue('account', 1, 10);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.setValue('account', 1, 10);
	await vivaldiWrapper.setValue('account', 1, 10, {
		gasLimit: 20000,
		gasPrice: 2000000000,
		nonce: 40
	});
	expect((vivaldiWrapper.contract.methods.setValue as jest.Mock).mock.calls).toMatchSnapshot();
	expect(setValueSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startRound', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	try {
		await vivaldiWrapper.startRound('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}

	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.startRound('account');
	await vivaldiWrapper.startRound('account', {
		gasLimit: 20000,
		gasPrice: 2000000000,
		nonce: 40
	});
	expect((vivaldiWrapper.contract.methods.startRound as jest.Mock).mock.calls).toMatchSnapshot();
	expect(startRoundSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('endRound', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	try {
		await vivaldiWrapper.endRound('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}

	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.endRound('account');
	await vivaldiWrapper.endRound('account', {
		gasLimit: 20000,
		gasPrice: 2000000000,
		nonce: 40
	});
	expect((vivaldiWrapper.contract.methods.endRound as jest.Mock).mock.calls).toMatchSnapshot();
	expect(endRoundSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('forceEndRound', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	try {
		await vivaldiWrapper.forceEndRound('account', 123, 456);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}

	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.forceEndRound('account', 123, 456);
	await vivaldiWrapper.forceEndRound('account', 123, 456, {
		gasLimit: 20000,
		gasPrice: 2000000000,
		nonce: 40
	});
	expect((vivaldiWrapper.contract.methods.forceEndRound as jest.Mock).mock.calls).toMatchSnapshot();
	expect(forceEndRoundSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('create', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.create('account', 1);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	expect(await vivaldiWrapper.create('account', 1)).toMatchSnapshot();
	expect(
		await vivaldiWrapper.create('account', 1, {
			gasPrice: 1000000000,
			gasLimit: 20000,
			nonce: 10
		})
	).toMatchSnapshot();
	expect(vivaldiWrapper.contract.methods.create as jest.Mock).toBeCalledTimes(2);
	expect(createSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('redeem', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.redeem('account', 1, 1);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	expect(await vivaldiWrapper.redeem('account', 1, 1)).toMatchSnapshot();
	expect(
		await vivaldiWrapper.redeem('account', 1, 1, {
			gasPrice: 1000000000,
			gasLimit: 20000,
			nonce: 10
		})
	).toMatchSnapshot();

	expect((vivaldiWrapper.contract.methods.redeem as jest.Mock).mock.calls).toMatchSnapshot();
	expect(redeemSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startCustodian', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.startCustodian(
			'account',
			'aAddr',
			'bAddr',
			'oracleAddr',
			123,
			true,
			true
		);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.startCustodian('account', 'aAddr', 'bAddr', 'oracleAddr', 123, true, true);
	await vivaldiWrapper.startCustodian(
		'account',
		'aAddr',
		'bAddr',
		'oracleAddr',
		123,
		true,
		true,
		{
			gasPrice: 1000000000,
			gasLimit: 20000,
			nonce: 10
		}
	);
	expect(
		(vivaldiWrapper.contract.methods.startCustodian as jest.Mock).mock.calls
	).toMatchSnapshot();
	expect(startCustodianSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('triggerPreReset', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.triggerPreReset('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.triggerPreReset('account');
	await vivaldiWrapper.triggerPreReset('account', {
		gasLimit: 20000,
		gasPrice: 1000000000,
		nonce: 30
	});
	expect(vivaldiWrapper.contract.methods.startPreReset as jest.Mock).toBeCalledTimes(2);
	expect(startPreResetSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('triggerReset', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.triggerReset('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.triggerReset('account');
	await vivaldiWrapper.triggerReset('account', {
		gasLimit: 20000,
		gasPrice: 1000000000,
		nonce: 40
	});
	expect(vivaldiWrapper.contract.methods.startReset as jest.Mock).toBeCalledTimes(2);
	expect(startResetSend.mock.calls).toMatchSnapshot();
	expect((web3Wrapper.getTransactionOption as jest.Mock).mock.calls).toMatchSnapshot();
});

test('updateRoleManager', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.updateRoleManager('account', 'newRoleManager');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.updateRoleManager('account', 'newRoleManager');
	await vivaldiWrapper.updateRoleManager('account', 'newRoleManager', {
		gasPrice: 2000000000,
		gasLimit: 200000,
		nonce: 10
	});
	expect(
		(vivaldiWrapper.contract.methods.updateRoleManager as jest.Mock).mock.calls
	).toMatchSnapshot();
	expect(
		(vivaldiWrapper.web3Wrapper.getTransactionOption as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('updateOperator', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.updateOperator('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.updateOperator('account');
	await vivaldiWrapper.updateOperator('account', {
		gasPrice: 2000000000,
		gasLimit: 200000,
		nonce: 10
	});
	expect(
		(vivaldiWrapper.web3Wrapper.getTransactionOption as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('updateOracle', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.updateOracle('account', 'newOracleAddress');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.updateOracle('account', 'newOracleAddress');
	await vivaldiWrapper.updateOracle('account', 'newOracleAddress', {
		gasPrice: 2000000000,
		gasLimit: 200000,
		nonce: 10
	});
	expect(
		(vivaldiWrapper.contract.methods.updateOracle as jest.Mock).mock.calls
	).toMatchSnapshot();
	expect(
		(vivaldiWrapper.web3Wrapper.getTransactionOption as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('updateFeeCollector', async () => {
	web3Wrapper.getTransactionOption = jest.fn(
		(account: string, gasLimit: number, option: ITransactionOption = {}) =>
			Promise.resolve({
				from: account,
				gasPrice: option.gasPrice || 1000000000,
				gas: option.gasLimit || gasLimit,
				nonce: option.nonce || 10
			})
	);
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await vivaldiWrapper.updateFeeCollector('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await vivaldiWrapper.updateFeeCollector('account');
	await vivaldiWrapper.updateFeeCollector('account', {
		gasPrice: 2000000000,
		gasLimit: 200000,
		nonce: 10
	});
	expect(
		(vivaldiWrapper.web3Wrapper.getTransactionOption as jest.Mock).mock.calls
	).toMatchSnapshot();
});
