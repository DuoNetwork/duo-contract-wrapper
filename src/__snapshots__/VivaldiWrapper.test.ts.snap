// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectFee 1`] = `"Read Only Mode"`;

exports[`collectFee 2`] = `
Array [
  Array [
    "10000000000000000000",
  ],
  Array [
    "10000000000000000000",
  ],
]
`;

exports[`collectFee 3`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 40000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`collectFee 4`] = `
Array [
  Array [
    "account",
    40000,
    Object {},
  ],
  Array [
    "account",
    40000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`create 1`] = `"Read Only Mode"`;

exports[`create 2`] = `"createTxHash"`;

exports[`create 3`] = `"createTxHash"`;

exports[`create 4`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 250000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`create 5`] = `
Array [
  Array [
    "account",
    250000,
    Object {},
  ],
  Array [
    "account",
    250000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`endRound 1`] = `"Read Only Mode"`;

exports[`endRound 2`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`endRound 3`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 300000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`endRound 4`] = `
Array [
  Array [
    "account",
    300000,
    Object {},
  ],
  Array [
    "account",
    300000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`forceEndRound 1`] = `"Read Only Mode"`;

exports[`forceEndRound 2`] = `
Array [
  Array [
    "123000000000000000000",
    456,
  ],
  Array [
    "123000000000000000000",
    456,
  ],
]
`;

exports[`forceEndRound 3`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 300000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`forceEndRound 4`] = `
Array [
  Array [
    "account",
    300000,
    Object {},
  ],
  Array [
    "account",
    300000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`getAddress 1`] = `
Object {
  "aToken": "aToken",
  "bToken": "bToken",
  "collateralToken": "collateralToken",
  "feeCollector": "feeCollector",
  "operator": "operator",
  "oracle": "oracle",
  "roleManager": "roleManager",
}
`;

exports[`getStates 1`] = `
Object {
  "clearCommRate": 100000000000000,
  "collateral": 1,
  "createCommRate": 1.0000000000000001e+21,
  "feeBalance": 1000,
  "isKnockedIn": true,
  "iterationGasThreshold": 100000000000000000000,
  "lastOperationTime": 12345678980000,
  "lastPrice": 2,
  "lastPriceTime": 1e+23,
  "maturity": 1e+21,
  "minBalance": 100000,
  "nextResetAddrIndex": 1e+21,
  "operationCoolDown": 1e+24,
  "period": 1e+21,
  "preResetWaitingBlocks": 1000000000000000000,
  "priceFetchCoolDown": 1e+21,
  "redeemCommRate": 100000000000000,
  "resetPrice": 1e-9,
  "resetPriceTime": 1e+23,
  "roundStrike": 1,
  "state": "Loading",
  "totalSupplyA": 1e-13,
  "totalSupplyB": 1e-16,
  "totalUsers": 10000000000000000000,
}
`;

exports[`getUserAddress 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`redeem 1`] = `"Read Only Mode"`;

exports[`redeem 2`] = `"redeemTxHash"`;

exports[`redeem 3`] = `"redeemTxHash"`;

exports[`redeem 4`] = `
Array [
  Array [
    "1000000000000000000",
    "1000000000000000000",
  ],
  Array [
    "1000000000000000000",
    "1000000000000000000",
  ],
]
`;

exports[`redeem 5`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 160000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`redeem 6`] = `
Array [
  Array [
    "account",
    160000,
    Object {},
  ],
  Array [
    "account",
    160000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`setValue 1`] = `"Read Only Mode"`;

exports[`setValue 2`] = `
Array [
  Array [
    1,
    10,
  ],
  Array [
    1,
    10,
  ],
]
`;

exports[`setValue 3`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 50000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`setValue 4`] = `
Array [
  Array [
    "account",
    50000,
    Object {},
  ],
  Array [
    "account",
    50000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`startCustodian 1`] = `"Read Only Mode"`;

exports[`startCustodian 2`] = `
Array [
  Array [
    "aAddr",
    "bAddr",
    "oracleAddr",
    "123000000000000000000",
    true,
    true,
  ],
  Array [
    "aAddr",
    "bAddr",
    "oracleAddr",
    "123000000000000000000",
    true,
    true,
  ],
]
`;

exports[`startCustodian 3`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 400000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`startCustodian 4`] = `
Array [
  Array [
    "account",
    400000,
    Object {},
  ],
  Array [
    "account",
    400000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`startRound 1`] = `"Read Only Mode"`;

exports[`startRound 2`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`startRound 3`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 300000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`startRound 4`] = `
Array [
  Array [
    "account",
    300000,
    Object {},
  ],
  Array [
    "account",
    300000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 2000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`triggerPreReset 1`] = `"Read Only Mode"`;

exports[`triggerPreReset 2`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 240000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 1000000000,
      "nonce": 30,
    },
  ],
]
`;

exports[`triggerPreReset 3`] = `
Array [
  Array [
    "account",
    240000,
    Object {},
  ],
  Array [
    "account",
    240000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 1000000000,
      "nonce": 30,
    },
  ],
]
`;

exports[`triggerReset 1`] = `"Read Only Mode"`;

exports[`triggerReset 2`] = `
Array [
  Array [
    Object {
      "from": "account",
      "gas": 4000000,
      "gasPrice": 1000000000,
      "nonce": 10,
    },
  ],
  Array [
    Object {
      "from": "account",
      "gas": 20000,
      "gasPrice": 1000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`triggerReset 3`] = `
Array [
  Array [
    "account",
    4000000,
    Object {},
  ],
  Array [
    "account",
    4000000,
    Object {
      "gasLimit": 20000,
      "gasPrice": 1000000000,
      "nonce": 40,
    },
  ],
]
`;

exports[`updateFeeCollector 1`] = `"Read Only Mode"`;

exports[`updateFeeCollector 2`] = `
Array [
  Array [
    "account",
    1000000,
    Object {},
  ],
  Array [
    "account",
    1000000,
    Object {
      "gasLimit": 200000,
      "gasPrice": 2000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`updateOperator 1`] = `"Read Only Mode"`;

exports[`updateOperator 2`] = `
Array [
  Array [
    "account",
    1000000,
    Object {},
  ],
  Array [
    "account",
    1000000,
    Object {
      "gasLimit": 200000,
      "gasPrice": 2000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`updateOracle 1`] = `"Read Only Mode"`;

exports[`updateOracle 2`] = `
Array [
  Array [
    "newOracleAddress",
  ],
  Array [
    "newOracleAddress",
  ],
]
`;

exports[`updateOracle 3`] = `
Array [
  Array [
    "account",
    1000000,
    Object {},
  ],
  Array [
    "account",
    1000000,
    Object {
      "gasLimit": 200000,
      "gasPrice": 2000000000,
      "nonce": 10,
    },
  ],
]
`;

exports[`updateRoleManager 1`] = `"Read Only Mode"`;

exports[`updateRoleManager 2`] = `
Array [
  Array [
    "newRoleManager",
  ],
  Array [
    "newRoleManager",
  ],
]
`;

exports[`updateRoleManager 3`] = `
Array [
  Array [
    "account",
    1000000,
    Object {},
  ],
  Array [
    "account",
    1000000,
    Object {
      "gasLimit": 200000,
      "gasPrice": 2000000000,
      "nonce": 10,
    },
  ],
]
`;
